File: tournaments/countClouds/countClouds.js
function countClouds(skyMap) {
    const markVisited = function() {
        while (store.length) {
            const [i, j] = store.shift();
            skyMap[i][j] = 'X';

            if (i + 1 in skyMap && '1' === skyMap[i + 1][j]) {
                store.push([i + 1, j]);
            }

            if (i - 1 in skyMap && '1' === skyMap[i - 1][j]) {
                store.push([i-1, j]);
            }

            if (j + 1 in skyMap[0] && '1' === skyMap[i][j + 1]) {
                store.push([i, j + 1]);
            }

            if (j - 1 in skyMap[0] && '1' === skyMap[i][j - 1]) {
                store.push([i, j - 1]);
            }
        }
    };

    if (!skyMap || !skyMap.length) {
        return 0;
    }

    const store = [];
    let result = 0;

    for (let i = 0; i in skyMap; ++i) {
        for (let j = 0; j in skyMap[0]; ++j) {
            if ('1' === skyMap[i][j]) {
                ++result;
                store.push([i, j]);
                markVisited();
            }
        }
    }

    return result;
}

File: tournaments/countClouds/countClouds.cs
int countClouds(char[][] skyMap) {
    if (null == skyMap || 0 == skyMap.Length) {
        return 0;
    }

    var connectedComponentsCount = 0;
    var rows = skyMap.Length;
    var cols = skyMap[0].Length;
    var dsu = new Dsu(1 + rows * cols);

    for (var row = 0; rows > row; ++row) {
        for (var col = 0; cols > col; ++col) {
            if ('1' == skyMap[row][col]) {
                if (rows > 1 + row && '1' == skyMap[1 + row][col]) {
                    dsu.union(cols * (1 + row) + col, cols * row + col);
                }
                if (cols > 1 + col && '1' == skyMap[row][1 + col]) {
                    dsu.union(cols * row + col + 1, cols * row + col);
                }
            } else {
                dsu.union(cols * row + col, rows * cols);
            }
        }
    }

    for (var i = 0; rows * cols > i; ++i) {
        if (i == dsu.find(i)) {
            ++connectedComponentsCount;
        }
    }

    return connectedComponentsCount;
}

private class Dsu {
    private readonly int[] _parents;
    private readonly int[] _ranks;

    public Dsu(int n) {
        _parents = new int[n];
        _ranks = new int[n];

        for (var i = 0; n > i; ++i) {
            _parents[i] = i;
        }
    }

    public int find(int x) {
        if (_parents[x] != x) {
            _parents[x] = find(_parents[x]);
        }

        return _parents[x];
    }

    public void union(int x, int y) {
        var px = find(x);
        var py = find(y);

        if (px != py) {
            if (_ranks[px] > _ranks[py]) {
                _parents[py] = px;
                ++_ranks[px];
            } else {
                _parents[px] = py;
                ++_ranks[py];
            }
        }
    }
}
